`send` - main page

thedoublejay opened the pull request:
<!--  Thanks for sending a pull request! -->

#### What kind of PR is this?:
<!-- Use one of the following kinds:
/kind feature
/kind fix
/kind chore
/kind docs
/kind refactor
/kind dependencies
-->

/kind feature

#### What this PR does / why we need it:
1. Balances page -> Click -> Links to Send page
2. Send Page
3. QR Code scanner
4. Form validation
   - Amount only accepts numbers with decimals
   - Both address and amount are required
5. `PrimaryButton` component that has ability to have disabled effect
6. `MAX` button that sets amount to token amount

Pending items:
1. Unit tests
2. E2E tests
3. Will merge submit logic once this is merged to main

#### Which issue(s) does this PR fixes?:
<!--
(Optional) Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #55 and #208

#### Additional comments?:
<div>
<img width="30%" src="https://user-images.githubusercontent.com/25013382/123395494-d8147580-d5d2-11eb-92fb-5cf0c9bcf249.jpg">
<img width="30%" src="https://user-images.githubusercontent.com/25013382/123395490-d64ab200-d5d2-11eb-802c-4d813d928f19.jpg">
<img width="30%" src="https://user-images.githubusercontent.com/25013382/123395491-d77bdf00-d5d2-11eb-8938-40e1cc6e0bb4.jpg">
<img width="30%" src="https://user-images.githubusercontent.com/25013382/123395495-d8ad0c00-d5d2-11eb-8c84-c602528b78f8.jpg">
<img width="30%" src="https://user-images.githubusercontent.com/25013382/123395808-30e40e00-d5d3-11eb-8f18-cd12a7bfac3f.jpg">
</div>

ivan-zynesis commented on screens/AppNavigator/screens/BalancesScreen/SendScreen/SendScreen.tsx:
```diff
@@ -0,0 +1,206 @@
+          await trigger('amount')
+        }}
+      />
+      <PrimaryButton disabled={!isValid} title='Send' onPress={onSubmit}>
```
Not familiar with this `useForm` API, but I cant identify what's the logic that decide whether it `isValid` here

thedoublejay replied:
@ivan-zynesis `useForm` API is from here - https://www.react-hook-form.com/. Basically, it's a hook to do form validations, like required, max amount and number pattern.

ivan-zynesis commented on screens/AppNavigator/screens/BalancesScreen/SendScreen/SendScreen.tsx:
```diff
@@ -0,0 +1,206 @@
+    <>
+      <Text
+        style={tailwind('text-sm text-gray-300 font-bold pl-4 mt-4 mb-2')}
+      >{translate('screens/SendScreen', 'TO ADDRESS')}
```
I think this text should be in black color

thedoublejay replied:
thanks! will check

ivan-zynesis commented on screens/AppNavigator/screens/BalancesScreen/SendScreen/SendScreen.tsx:
```diff
@@ -0,0 +1,206 @@
+  const { control, setValue, formState: { isValid }, getValues, trigger } = useForm({ mode: 'onChange' })
+
+  const onSubmit = (): void => {
+    console.log(isValid)
```
Will come back later after this part updated

fuxingloh commented:
![screenshot](https://netlify-cocoon.netlify.app/.netlify/functions/fetch?code=307&path=eyJzaXRlX2lkIjoiMDM3YmVkYzUtYTM2Ni00NDBlLThlOGYtZmE4YmViNjY3NjlmIiwiZGVwbG95X2lkIjoiNjBkOTNhZTExZmYxODEwMDA3NDhhNzc4IiwiaWQiOiJlMzFiMzkwNC01MDVhLTQ5YWYtYjVkYi00ZGRlYWMyMzk2OTQifQ==)

DFI Token cannot be send, only DFI UTXO can be send

<details><summary>Browser metadata</summary>





```
Path:      /app/Balances
Browser:   Chrome 91.0.4472.114 on Mac OS 10.15.7
Viewport:  1680 x 971 @2x
Language:  en-GB
Cookies:   Enabled
```

[View on BrowserStack](https://live.browserstack.com/dashboard#url=https%3A%2F%2Fdeploy-preview-202--defi-wallet.netlify.app%2Fapp%2FBalances&os=OS+X&os_version=Catalina&browser=Chrome&browser_version=91.0&resolution=1680x971&speed=1&start=true)



</details>

[Open Deploy Preview](https://deploy-preview-202--defi-wallet.netlify.app/app/Balances) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=60d93ae11ff181000748a778&commentID=60dac83be98390680e9b919f&resolution=resolved)

fuxingloh commented on components/index.tsx:
```diff
@@ -35,3 +45,17 @@ export function NumberText (props: { value: string, decimal?: number, style?: St
   )
 }
+
+export function PrimaryButton (props: React.PropsWithChildren<ButtonProps>): JSX.Element {
+  return (
+    <TouchableOpacity
+      disabled={props.disabled}
+      onPress={props.onPress}
+      style={[tailwind('m-4 mt-8 p-3 rounded flex-row justify-center'), props.disabled === true ? DisabledColorStyle.button : PrimaryColorStyle.button]}
+    >
+      {
+        props.children
+      }
+    </TouchableOpacity>
+  )
+}
```
should move this to an individual file

we need to start maintaining them as` components/*.tsx`

fuxingloh commented on constants/Theme.ts:
```diff
@@ -15,6 +15,17 @@ export const PrimaryColorStyle = StyleSheet.create({
+  }
+})
+
+export const DisabledColorStyle = StyleSheet.create({
+  button: {
+    backgroundColor: 'rgba(0,0,0,0.2)',
+    color: 'rgba(255,255,255,0.06)'
```
move this within the component as it should be scoped within that scope only

fuxingloh commented on constants/Theme.ts:
```diff
@@ -15,6 +15,17 @@ export const PrimaryColorStyle = StyleSheet.create({
   border: {
     borderColor: PrimaryColor
+  },
+  button: {
+    backgroundColor: PrimaryColor,
+    color: '#ffffff'
+  }
```
don't use feature style, should only declare utility style

fuxingloh replied:
aka more into component level

thedoublejay replied:
ok copy

fuxingloh commented on screens/AppNavigator/screens/BalancesScreen/SendScreen/SendScreen.tsx:
```diff
@@ -0,0 +1,206 @@
+import { translate } from '../../../../../translations'
+import { BalanceParamList } from '../BalancesNavigator'
+
+interface QRScannerProps {
+  onScan: (code: string) => void
+  isQRScanning: boolean
+  onCloseModal: () => void
+}
```
Move all QR features into another PR, don't prioritize this yet. We will track that individually.

thedoublejay replied:
Ok, do you want me to remove them here? It's already working well so I'm not sure what's the point of moving it out

fuxingloh reviewed (changes requested):
> initial review

- move QR scanner into another Pr
- DFI UTXO can be send
- DFI Token cannot be send

thedoublejay commented:
@fuxingloh I have made some changes:
1. Disabled the click for DFI token
2. Removed usage of `NumberText` 
3. Moved PrimaryButton to dedicated component

I have left a comment for the QR scanner part. For me, I think we can leave it since it's already working well and it's part of the implementation. Let me know your thoughts about it

fuxingloh commented:
> @fuxingloh I have made some changes:
> 
> 1. Disabled the click for DFI token
> 2. Removed usage of `NumberText`
> 3. Moved PrimaryButton to dedicated component
> 
> I have left a comment for the QR scanner part. For me, I think we can leave it since it's already working well and it's part of the implementation. Let me know your thoughts about it

Best to move it out, it's not maintainable for iterative reasons. Non-maintainable being: QR scanner has Modal component which is a design anti-pattern as it's easy to get wrong (TLDR don't use Modal in mobile). As this should be prioritized for merge, removing it for another PR will expedite the process.

thedoublejay commented:
@fuxingloh ok I have removed the QR code changes
